File: python-tiny/lexer.py
from ast import Set
import enum
import sys

class Lexer:

  valid_strings = set([])

  def __init__(self, input):
    self.source = input + '\n'
    self.curChar = ''
    self.curPos = -1
    self.nextChar()

  # process next char
  def nextChar(self):
    self.curPos += 1
    if self.curPos >= len(self.source):
      self.curChar = '\0'
    else:
      self.curChar = self.source[self.curPos]

  # return the lookahead char
  def peek(self):
    if self.curPos + 1 >= len(self.source):
      return '\0'
      
    return self.source[self.curPos + 1]

  # invalid token found, print error and exit
    # Invalid token found, print error message and exit.
  def abort(self, message):
      sys.exit("Lexing error. " + message)
    

  # skip whitespace except newlines, which we will use to indicate end of statement
  def skipWhitespace(self):
    while self.curChar == ' ' or self.curChar == '\t' or self.curChar == '\r':
        self.nextChar()

  # skip comments in the code
  def skipComments(self):
    if self.curChar == '#':
        while self.curChar != '\n':
            self.nextChar()

  # return in next token
  def getToken(self):
    self.skipWhitespace()
    self.skipComments()
    token = None

    # Check the first character of this token to see if we can decide what it is.
    # If it is a multiple character operator (e.g., !=), number, identifier, or keyword then we will process the rest.
    if self.curChar == '+':
        token = Token(self.curChar, TokenType.PLUS)
    elif self.curChar == '-':
        token = Token(self.curChar, TokenType.MINUS)
    elif self.curChar == '*':
        token = Token(self.curChar, TokenType.ASTERISK)
    elif self.curChar == '/':
        token = Token(self.curChar, TokenType.SLASH)
    elif self.curChar == '\n':
        token = Token(self.curChar, TokenType.NEWLINE)
    elif self.curChar == '\0':
        token = Token('', TokenType.EOF)
    elif self.curChar == '=':
        # Check whether this token is = or ==
        if self.peek() == '=':
            lastChar = self.curChar
            self.nextChar()
            token = Token(lastChar + self.curChar, TokenType.EQEQ)
        else:
            token = Token(self.curChar, TokenType.EQ)
    elif self.curChar == '>':
        # Check whether this is token is > or >=
        if self.peek() == '=':
            lastChar = self.curChar
            self.nextChar()
            token = Token(lastChar + self.curChar, TokenType.GTEQ)
        else:
            token = Token(self.curChar, TokenType.GT)
    elif self.curChar == '<':
            # Check whether this is token is < or <=
            if self.peek() == '=':
                lastChar = self.curChar
                self.nextChar()
                token = Token(lastChar + self.curChar, TokenType.LTEQ)
            else:
                token = Token(self.curChar, TokenType.LT)
    elif self.curChar == '!':
        if self.peek() == '=':
            lastChar = self.curChar
            self.nextChar()
            token = Token(lastChar + self.curChar, TokenType.NOTEQ)
        else:
            self.abort("Expected !=, got !" + self.peek())
    elif self.curChar == '\"':
        # Get characters between quotations.
        self.nextChar()
        startPos = self.curPos

        while self.curChar != '\"':
            # Don't allow special characters in the string. No escape characters, newlines, tabs, or %.
            # We will be using C's printf on this string.
            if self.curChar == '\r' or self.curChar == '\n' or self.curChar == '\t' or self.curChar == '\\' or self.curChar == '%':
                self.abort("Illegal character in string.")
            self.nextChar()

        tokText = self.source[startPos : self.curPos] # Get the substring.
        token = Token(tokText, TokenType.STRING)
    else:
        # Unknown token!
        self.abort("Unknown token: " + self.curChar)
  
    self.nextChar()
    return token

# Token contains the original text and the type of token.
class Token:   
  def __init__(self, tokenText, tokenKind):
      self.text = tokenText   # The token's actual text. Used for identifiers, strings, and numbers.
      self.kind = tokenKind   # The TokenType that this token is classified as.

# TokenType is our enum for all the types of tokens.
class TokenType(enum.Enum):
	EOF = -1
	NEWLINE = 0
	NUMBER = 1
	IDENT = 2
	STRING = 3
	# Keywords.
	LABEL = 101
	GOTO = 102
	PRINT = 103
	INPUT = 104
	LET = 105
	IF = 106
	THEN = 107
	ENDIF = 108
	WHILE = 109
	REPEAT = 110
	ENDWHILE = 111
	# Operators.
	EQ = 201  
	PLUS = 202
	MINUS = 203
	ASTERISK = 204
	SLASH = 205
	EQEQ = 206
	NOTEQ = 207
	LT = 208
	LTEQ = 209
	GT = 210
	GTEQ = 211

File: rustcompc/src/lexer.rs
use std::{fmt, default};


// Define our error types. These may be customized for our error handling cases.
// Now we will be able to write our own errors, defer to an underlying error
// implementation, or do something in between.
#[derive(Debug, Clone)]
struct LexerError {
  msg: String
}

type Result<'a, T> = std::result::Result<T, LexerError>;

// Generation of an error is completely separate from how it is displayed.
// There's no need to be concerned about cluttering complex logic with the display style.
//
// Note that we don't store any extra info about the errors. This means we can't state
// which string failed to parse without modifying our types to carry that information.
impl fmt::Display for LexerError {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
      write!(f, "Lexer Error: invalid token {}", self.msg)
  }
}

pub struct Lexer {
  source: String,
  pub curr_char: char,
  pub cur_pos: usize,
}

impl Lexer {
  pub fn build_lexer(input: String) -> Lexer {
    Lexer {
      source: input + "\0",
      curr_char: '\0',
      cur_pos: 0
    }
  }

  pub fn next_char(&mut self) -> char {
    if self.cur_pos >= self.source.len() {
      self.curr_char = '\0';
    } else {
      self.cur_pos =  self.cur_pos + 1;
      let byte = self.source.as_bytes()[self.cur_pos];
      self.curr_char = byte as char;
    }
    self.curr_char
  }

  pub fn peek(&mut self) -> char {
    if self.cur_pos + 1 >= self.source.len() {
      return '\0'
    }

    let byte = self.source.as_bytes()[self.cur_pos + 1];
    self.curr_char = byte as char;
    return self.curr_char
  }
  
  fn skip_whitespace(&mut self) {
    while self.curr_char == ' ' || self.curr_char == '\t' || self.curr_char == '\r' {
      self.next_char();
    }
  }

  fn skip_comments(&mut self) {
    if self.curr_char == '#' {
      while self.curr_char != '\n' {
        self.next_char();
      }
    }
  }

  pub fn get_token(&mut self) -> Token {
    self.skip_comments();
    self.skip_whitespace();
    let token: Result<Token> = match self.next_char() {
      '+' => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::PLUS }),
      '-' => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::MINUS }),
      '*' => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::ASTERISK }),
      '/' => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::SLASH }),
      '\n' => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::NEWLINE }),
      '\0' => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::EOF }),
      '=' => {
        match self.peek() {
            '=' => {
              self.next_char();
              Ok(Token { text: self.curr_char.to_string(), kind: TokenType::EQEQ})
            }
            _ => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::EQ }),
        }
      },
      '>' => {
        match self.peek() {
            '=' => {
              self.next_char();
              Ok(Token { text: self.curr_char.to_string(), kind: TokenType::GTEQ})
            }
            _ => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::GT }),
        }
      },
      '<' => {
        match self.peek() {
            '=' => {
              self.next_char();
              Ok(Token { text: self.curr_char.to_string(), kind: TokenType::LTEQ})
            }
            _ => Ok(Token { text: self.curr_char.to_string(), kind: TokenType::LT }),
        }
      },
      '!' => {
        match self.peek() {
            '=' => {
              self.next_char();
              Ok(Token { text: self.curr_char.to_string(), kind: TokenType::NOTEQ})
            }
            _ => Err(LexerError { msg: String::from("! not valid without =") }),
        }
      },
      '\"' => {
        self.next_char();
        let start_pos = self.cur_pos;

        while self.curr_char != '\"' {
          let val = match self.curr_char {
            '\r' => Err(LexerError { msg: String::from("invalid tab character")}),
            '\n' => Err(LexerError { msg: String::from("invalid charridge character")}),
            '\t' => Err(LexerError { msg: String::from("invalid tab character")}),
            '\\' => Err(LexerError { msg: String::from("invalid slash character")}),
            '%' => Err(LexerError { msg: String::from("invalid percentage character")}),
            _default => {
              self.next_char();
              Ok(self.curr_char)
            },
          };
          val.unwrap();
        };
        let tok_text = self.source.get_mut(start_pos..self.cur_pos).unwrap();
        Ok(Token { text: String::from(tok_text), kind: TokenType::STRING })
      }
      _ => Err(LexerError { msg: String::from("unknown token") }),
    };
    self.next_char();
    token.unwrap()
  } 
}

pub struct Token {
  pub text: String,
  pub kind: TokenType,
}

pub enum TokenType {
  EOF = -1,
	NEWLINE = 0,
	NUMBER = 1,
	IDENT = 2,
	STRING = 3,
	// Keywords.
	LABEL = 101,
	GOTO = 102,
	PRINT = 103,
	INPUT = 104,
	LET = 105,
	IF = 106,
	THEN = 107,
	ENDIF = 108,
	WHILE = 109,
	REPEAT = 110,
	ENDWHILE = 111,
	// Operators.
	EQ = 201,
	PLUS = 202,
	MINUS = 203,
	ASTERISK = 204,
	SLASH = 205,
	EQEQ = 206,
	NOTEQ = 207,
	LT = 208,
	LTEQ = 209,
	GT = 210,
	GTEQ = 211
}
